### PR TITLE
Add SignalWaiter utility class

### PR DESCRIPTION
**Pull request by bocon13**
This class is initialized statically on program start and registers handlers to listen for SIGINT, SIGTERM, and SIGUSR2.

Calls to WaitForSignal()  block until one of the signals is received.

Users can also call HandleSignal() to deliver a signal to waiters instead of kill() or pthread_kill().

TODO
- [ ] Consider unlocking sem_ in destructor
- [ ] Add signal callback register function that automatically spins up in a thread
- [ ] Should we detach threads when spinning up a waiter? Cleaning them up would happen automatically when they return.

### CODE REVIEW

File: stratum/lib/signal_waiter.cc [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    old_signal_handlers_[s] = h;
+  }
+}
+
+SignalWaiter::~SignalWaiter() {
+  // Register the old handlers for all the signals.
+  for (const auto& e : old_signal_handlers_) {
+    signal(e.first, e.second);
+  }
+  old_signal_handlers_.clear();
```
Comment (bocon13) [inline stratum/lib/signal_waiter.cc:42]: About to be destroyed, so this is superfluous.
```suggestion
```

File: stratum/lib/signal_waiter_test.cc [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+  }
+
+  void TearDown() override {
+    int sem_value = GetSemaphoreValue();
+    EXPECT_LE(1, sem_value);
+    LOG(INFO) << "Ended test with semaphore value: " << sem_value;
+    Reset();
+  }
+
+  void Reset() {
```
Comment (bocon13) [inline stratum/lib/signal_waiter_test.cc:35]:
```suggestion
    ResetSem();
  }

  void ResetSem() {
```

**Comment by bocon13** [comment]
Coverage should be 100%, but it looks like we don't get coverage data inside ASSERT_DEATH calls.

```
TEST_F(SignalWaiterTest, ConstructorError) {
  ::std::string error_msg("Failed to register signal: ");
  error_msg += strsignal(SIGKILL);
  ASSERT_DEATH(ConstructWaiter({SIGKILL}), error_msg);

  // Send a registered signal because the test framework expects one.
  EXPECT_OK(SignalWaiter::HandleSignal(SIGINT));
}
```

File: stratum/lib/signal_waiter_test.cc [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+  EXPECT_EQ(sa1.sa_handler, sa2.sa_handler);
+
+  // Send a registered signal because the test framework expects one.
+  EXPECT_OK(SignalWaiter::HandleSignal(SIGINT));
+}
+
+TEST_F(SignalWaiterTest, ConstructorError) {
+  ::std::string error_msg("Failed to register signal: ");
+  error_msg += strsignal(SIGKILL);
+  ASSERT_DEATH(ConstructWaiter({SIGKILL}), error_msg);
```
Comment (bocon13) [inline stratum/lib/signal_waiter_test.cc:238]:
```suggestion
  EXPECT_DEATH(ConstructWaiter({SIGKILL}), error_msg);
```

File: stratum/lib/signal_waiter_test.cc [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+  }
+
+  void TearDown() override {
+    int sem_value = GetSemaphoreValue();
+    EXPECT_LE(1, sem_value);
+    LOG(INFO) << "Ended test with semaphore value: " << sem_value;
+    ResetSem();
+  }
+
+  void ResetSem() {
```
Comment (bocon13) [inline stratum/lib/signal_waiter_test.cc:35]:
```suggestion
    ResetSemaphore();
  }

  void ResetSemaphore() {
```

File: stratum/lib/signal_waiter.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  // Blocking call to wait for one of the signals. Returns when one of the
+  // signals is received.
+  static void WaitForSignal();
+
+  // SignalWaiter is neither copyable nor movable.
+  SignalWaiter(const SignalWaiter&) = delete;
+  SignalWaiter& operator=(const SignalWaiter&) = delete;
+
+  // Static instance that gets created at program start.
+  static SignalWaiter instance_;
```
Comment (bocon13) [inline stratum/lib/signal_waiter.h:40]: Make this static const

File: stratum/lib/signal_waiter.h [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  ~SignalWaiter();
+
+  // Semaphore that is used to block threads that call WaitForSignal(). The
+  // semaphore is initialized locked (value 0). Waiting threads call sem_wait(),
+  // and the signal handler calls sem_post(). The semaphone remains unlocked
+  // (value >= 1) after a signal has been handled.
+  sem_t sem_;
+
+  // Vector of signals for which we registered handlers.
+  std::vector<int> signals_;
```
Comment (bocon13) [inline stratum/lib/signal_waiter.h:54]: const?

File: stratum/lib/signal_waiter.cc [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+// Signal received callback which is registered as the handler for signals using
+// signal() system call.
+void SignalRcvCallback(int value) { SignalWaiter::HandleSignal(value); }
+
+}  // namespace
+
+// Initialize static member.
+SignalWaiter SignalWaiter::instance_;
+
+SignalWaiter::SignalWaiter(const std::vector<int> signals) : signals_(signals) {
```
Comment (bocon13) [inline stratum/lib/signal_waiter.cc:24]: Make pattern, so that we can return a StatusOr

Comment (pudelkoM) [inline stratum/lib/signal_waiter.cc:24]: Consider set to prevent duplicate signals and weird behavior after deconstruction.

**Comment by pudelkoM** [comment]
After looking at this in more detail, some other points:

- This class is inherently not thread-safe and should say so in the docs

- Other classes must not use this class in their static initialization (ordering issue)

- If possible, the constructor should be as simple as possible; never throw or abort. 
   - Counter point: guaranteeing `sem_init` makes `WaitForSignal` safe to call independent of manual init.


Maybes:

- Explicit start and shutdown (`Activate()`, `Deactivate()`), instead of de/constructor, to signal clear usage. Could also make it more testable

- Decide if we want a barrier (unlock once) or a notification

- Is there a use case for having multiple threads waiting on `WaitForSignal`? Should we register functions instead?

File: stratum/lib/signal_waiter.cc [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  for (const auto& e : old_signal_handlers_) {
+    signal(e.first, e.second);
+  }
+
+  CHECK_ERR(sem_destroy(&sem_));
+}
+
+// Reminder: async-signal-safe function calls only!
+::util::Status SignalWaiter::HandleSignal(int value) {
+  const auto& s = instance_.signals_;
```
Comment (bocon13) [inline stratum/lib/signal_waiter.cc:48]: If we want to log, use RAW_LOG

https://github.com/google/glog/blob/master/src/glog/raw_logging.h.in

File: stratum/lib/signal_waiter.cc [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  if (std::find(s.begin(), s.end(), value) == s.end()) {
+    RETURN_ERROR(ERR_INTERNAL)
```
Comment (bocon13) [inline stratum/lib/signal_waiter.cc:50]: Check to make sure that this does not call new

Comment (pudelkoM) [reply stratum/lib/signal_waiter.cc:50]: I do not have concrete evidence, as that depends on the implementation, but there is this note on the docs:
> If the algorithm fails to allocate memory, std::bad_alloc is thrown. 

https://en.cppreference.com/w/cpp/algorithm/find


Maybe this is another plus of using a set: https://en.cppreference.com/w/cpp/container/unordered_set/count

**Comment by pudelkoM** [comment]
Check if we can use a function returning a pointer to static variable for the lock-free singleton requirement: https://youtu.be/xu7q8dGvuwk?t=2390

**Comment by bocon13** [comment]
That’s actually pretty clever. I’ll try to see if I can refactor to use
that pattern.

File: stratum/lib/signal_waiter_test.cc [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+  EXPECT_OK(SignalWaiter::HandleSignal(SIGINT));
+  SignalWaiter::WaitForSignal();
+  SignalWaiter::WaitForSignal();
+}
+
+TEST_F(SignalWaiterTest, DoubleSignalBeforeWait) {
+  EXPECT_OK(SignalWaiter::HandleSignal(SIGINT));
+  EXPECT_OK(SignalWaiter::HandleSignal(SIGINT));
+  SignalWaiter::WaitForSignal();
+  EXPECT_EQ(2, GetSemaphoreValue());
```
Comment (pudelkoM) [inline stratum/lib/signal_waiter_test.cc:66]: We should not test the internal state, only the public API.

**Comment by bocon13** [comment]
superseded by #652

**Comment by pudelkoM** [comment]
We can revisit this if the async shutdown problem re-surfaces in other binaries.

**Pull request closed without merging**